be/src/storage/tablet_updates.cpp

```diff
@@ -1804,8 +1806,188 @@ Status TabletUpdates::load_from_base_tablet(int64_t request_version, Tablet* bas
+    DCHECK(_tablet.tablet_state() == TABLET_NOTREADY)
+            << "perform_directly_schema_change is only allowed in schema change process";
+    LOG(INFO) << "perform_directly_schema_change start tablet:" << _tablet.tablet_id()
+              << " request_version:" << request_version << " #pending:" << _pending_commits.size();
+    int64_t max_version = base_tablet->updates()->max_version();
+    if (max_version < request_version) {
+        LOG(WARNING) << "perform_directly_schema_change base_tablet's max_version:" << max_version
+                     << " < alter_version:" << request_version << " tablet:" << _tablet.tablet_id()
+                     << " base_tablet:" << base_tablet->tablet_id();
+        return Status::InternalError("perform_directly_schema_change max_version < request_version");
+    }
+    std::vector<RowsetSharedPtr> rowsets;
+    EditVersion version;
+    Status status = base_tablet->updates()->get_applied_rowsets(request_version, &rowsets, &version);
+    if (!status.ok()) {
+        LOG(WARNING) << "perform_directly_schema_change get base tablet rowsets error tablet:"
+                     << base_tablet->tablet_id() << " request_version:" << request_version << " reason:" << status;
+        return status;
+    }
+
+    // disable compaction temporarily when tablet just loaded
+    _last_compaction_time_ms = UnixMillis();
+
+    auto kv_store = _tablet.data_dir()->get_meta();
+    auto tablet_id = _tablet.tablet_id();
+
+    // delete old meta
+    auto data_dir = _tablet.data_dir();
+    rocksdb::WriteBatch wb;
+    RETURN_IF_ERROR(TabletMetaManager::clear_log(data_dir, &wb, tablet_id));
+    RETURN_IF_ERROR(TabletMetaManager::clear_rowset(data_dir, &wb, tablet_id));
+    RETURN_IF_ERROR(TabletMetaManager::clear_pending_rowset(data_dir, &wb, tablet_id));
+    RETURN_IF_ERROR(TabletMetaManager::clear_del_vector(data_dir, &wb, tablet_id));
+
+    std::unique_lock wrlock(_tablet.get_header_lock());
+    status = kv_store->write_batch(&wb);
+    if (!status.ok()) {
+        LOG(WARNING) << "Fail to delete old meta and write new meta" << tablet_id << ": " << status;
+        return Status::InternalError("Fail to delete old meta and write new meta");
+    }
+
+    vectorized::Schema base_schema = vectorized::ChunkHelper::convert_schema_to_format_v2(base_tablet->tablet_schema());
+    vectorized::TabletReaderParams read_params;
+    read_params.reader_type = ReaderType::READER_ALTER_TABLE;
+    read_params.skip_aggregation = false;
+    read_params.chunk_size = config::vector_chunk_size;
+
+    for (const auto& rowset : rowsets) {
+        LOG(INFO) << "begin to convert a history rowset. version=" << rowset->version().first << "-"
+                  << rowset->version().second;
+
+        std::unique_ptr<vectorized::TabletReader> tablet_reader =
+                std::make_unique<vectorized::TabletReader>(base_tablet, Version(0, version.major()), base_schema);
```

Comment:
Why iterate all the rowsets and in each iteration read&write the whole tablet? The will write the whole tablet multiple times, and tablet_reader doesn't ensure the output is sorted by primary key, write the reader output to rowset_writer will corrupt the file.

Reply:
changed.